lib/ignored-paths.js

```diff
@@ -229,23 +243,54 @@ class IgnoredPaths {
      */
 
     addPatternRelativeToCwd(ig, pattern) {
-        ig.addPattern(
-            this.getBaseDir() === this.options.cwd
-                ? pattern
-                : relativize(pattern, path.relative(this.ignoreFileDir, this.options.cwd))
-        );
+        const baseDir = this.getBaseDir();
+        const cookedPattern = baseDir === this.options.cwd
+            ? pattern
+            : relativize(pattern, path.relative(baseDir, this.options.cwd));
+
+        ig.addPattern(cookedPattern);
+        debug("addPatternRelativeToCwd:\n  original = %s\n  cooked   = %s", pattern, cookedPattern);
     }
 
     addPatternRelativeToIgnoreFile(ig, pattern) {
-        ig.addPattern(
-            this.getBaseDir() === this.ignoreFileDir
-                ? pattern
-                : relativize(pattern, path.relative(this.options.cwd, this.ignoreFileDir))
-        );
+        const baseDir = this.getBaseDir();
+        const cookedPattern = baseDir === this.ignoreFileDir
+            ? pattern
+            : relativize(pattern, path.relative(baseDir, this.ignoreFileDir));
+
+        ig.addPattern(cookedPattern);
+        debug("addPatternRelativeToIgnoreFile:\n  original = %s\n  cooked   = %s", pattern, cookedPattern);
     }
 
+    // Detect the common ancestor
     getBaseDir() {
-        return this.ignoreFileDir.startsWith(this.options.cwd) ? this.options.cwd : this.ignoreFileDir;
+        if (!this._baseDir) {
+            const a = path.resolve(this.options.cwd);
+            const b = path.resolve(this.ignoreFileDir);
+            let lastSepPos = 0;
+
+            // Set the shorter one (it's the common ancestor if one includes the other).
+            this._baseDir = a.length < b.length ? a : b;
+
+            // Set the common ancestor.
+            for (let i = 0; i < a.length && i < b.length; ++i) {
```

Comment:
Would it work to split on `path.sep` rather than iterating over each character?

```js
const cwd = path.resolve(this.options.cwd);
const ignoreFileDir = path.resolve(this.ignoreFileDir);

if (cwd === ignoreFileDir) {
    this._baseDir = cwd;
} else {
    const cwdParts = cwd.split(path.sep);
    const ignoreFileDirParts = ignoreFileDir.split(path.sep);

    const longerParts = cwdParts.length > ignoreFileDirParts.length
        ? cwdParts
        : ignoreFileDirParts;

    this._baseDir = longerParts.slice(
        0,
        longerParts.findIndex((_, index) => cwdParts[index] !== ignoreFileDirParts[index])
    );
}
```

Reply:
Hm. I wrote that code as being careful to use only one traverse on the path strings. Indeed, though it's `O(n)` as is...